BoxContentSDK/BoxContentSDK/BOXNSURLSessionManager.h

```diff
@@ -10,21 +10,98 @@
 
 NS_ASSUME_NONNULL_BEGIN
 
+@protocol BOXNSURLSessionTaskDelegate <NSObject>
+
+/**
+ * To be called to finish the operation for a NSURLSessionTask upon its completion
+ *
+ * @param response  The response received from Box as a result of the API call.
+ * @param error     An error in the NSURLErrorDomain
+ */
+- (void)finishURLSessionTaskWithResponse:(NSURLResponse *)response error:(NSError *)error;
+
+@optional
+
+/**
+ * To be called to process the intermediate response from the task
+ *
+ * @param response  The intermediate response received from Box as a result of the API call
+ */
+- (void)processIntermediateResponse:(NSURLResponse *)response;
+
+/**
+ * To be called to process the intermediate data from the task
+ *
+ * @param data  The intermediate data received from Box as a result of the API call
+ */
+- (void)processIntermediateData:(NSData *)data;
+
+@end
+
+@protocol BOXNSURLSessionDownloadTaskDelegate <BOXNSURLSessionTaskDelegate>
+
+@optional
+
+/**
+ * Output stream to write data into. If destinationPath is provided, outputStream will be ignored
+ * Using outputStream to consume data will not allow the request to be executed in the background if the app is killed/suspended
+ */
+- (NSOutputStream *)outputStream;
+
+/**
+ * The location for output file. If provided, outputStream will be ignored
+ * Using destinationPath to consume data will allow request to be executed in the background if the app is killed/suspended and resume upon app restarts/resumes
+ */
+- (NSString *)destinationPath;
```

Comment:
What are the rules for this path? Do we expect the directory to exist, and the file not to? What happens if either of those are false?